R/course.R

```diff
@@ -0,0 +1,324 @@
+## see end of file for some cURL notes
+
+#' Download course materials
+#'
+#' Special-purpose function to download a folder of course materials. The only
+#' demand on the user is to confirm or specify where the new folder should be
+#' stored. Workflow:
+#' * User calls `use_course("SHORTLINK-GOES-HERE")`.
+#' * User is asked to notice and confirm the location of the new folder Specify
+#' `destdir` to skip this.
+#' * User is asked if they'd like to delete the ZIP file.
+#' * New folder is opened in the file manager, e.g. Finder or File Explorer.
+#'
+#' @param url Link to a ZIP file containing the materials, possibly behind a
+#'   shortlink. Function developed with DropBox and GitHub in mind, but should
+#'   work for ZIP files generally. See [download_zip()] for more.
+#' @param destdir The new folder is stored here. Defaults to working directory.
+#'
+#' @return Path to the new directory holding the course materials.
+#' @export
+#' @family download functions
+#' @examples
+#' \dontrun{
+#' ## demo with a small CRAN package available in various places
+#'
+#' ## from CRAN
+#' use_course("https://cran.r-project.org/bin/windows/contrib/3.4/rematch2_2.0.1.zip")
+#'
+#' ## from GitHub, 2 ways
+#' use_course("https://github.com/r-lib/rematch2/archive/master.zip")
+#' use_course("http://github.com/r-lib/rematch2/zipball/master/")
+#' }
+use_course <- function(url, destdir = NULL) {
+  zipfile <- download_zip(url, destdir = destdir, pedantic = TRUE)
+  tidy_unzip(zipfile)
+}
+
+#' Download a ZIP file
+#'
+#' Special-purpose function to download a ZIP file and automatically determine
+#' the file name, which often determines the folder name after unpacking.
+#' Developed for use in live teaching, with DropBox and GitHub as primary
+#' targets, possibly via shortlinks. Both platforms offer a way to download an
+#' entire folder or repo as a ZIP file, with information about the original
+#' folder or repo transmitted in the `Content-Disposition` header. In the
+#' absence of this header, a filename is generated from the input URL. In either
+#' case, the filename is sanitized.
+#'
+#' @section DropBox:
+#'
+#' To make a folder available for ZIP download, create a shared link for it:
+#' * <https://www.dropbox.com/help/files-folders/view-only-access>
+#'
+#' A shared link will have this form:
+#' ```
+#' https://www.dropbox.com/sh/12345abcde/6789wxyz?dl=0
+#' ```
+#' Replace the `dl=0` at the end with `dl=1` to create a download link. The ZIP
```

Comment:
Can we do this automatically?

Reply:
Do you mean purely mechanically, i.e. on the front-end?

If yes: I could special-case DropBox and check the input url and make this substitution.

Otherwise, I don't think so. If input is a shortlink, we have no visibility into the redirects, to get at this URL and fix it. The failure mode is also non-specific (`Error: Download does not have MIME type 'application/zip'`).

Reply:
I have now learned this:

> Information about any short bitly URL `http://bit.ly/x` is available at `http://bit.ly/x+` (that is, the URL with a plus sign appended), for example `http://bit.ly/1sNZMwL+`. This allows users to see and check the long URL before visiting it.

It's not designed for programmatic use but maybe helpful. In any case, probably not doing this now.